wandb/sdk/launch/agent2/controllers/base.py

```diff
@@ -157,3 +193,146 @@ async def find_orphaned_jobs(self) -> List[Any]:
+        )
+        fnames = None
+        phase = None
+        if tracker is not None:
+            phase = tracker.err_stage
+            fnames = tracker.saver.save_contents("".join(tb_str), "error.log", "error")
+        else:
+            phase = "agent"
+
+        fail_run_queue_item = event_loop_thread_exec(
+            self.jobset.api.fail_run_queue_item
+        )
+        await fail_run_queue_item(item_id, str(exception), phase, fnames)
+
+        return
+
+    async def fail_unsubmitted_run(self, item_id: str) -> None:
+        fail_run_queue_item = event_loop_thread_exec(
+            self.jobset.api.fail_run_queue_item
+        )
+        await fail_run_queue_item(
+            item_id,
+            "The job was not submitted successfully",
+            "agent",
+        )
+        return
+
+
+async def check_run_called_init(
+    api: Api,
+    run: AbstractRun,
+    entity: str,
+    project: str,
+    run_id: str,
+    run_queue_item_id: str,
+) -> Tuple[bool, Optional[str]]:
+    called_init = False
+    # We do some weird stuff here getting run info to check for a
+    # created in run in W&B.
+    #
+    # We retry for 60 seconds with an exponential backoff in case
+    # upsert run is taking a while.
+    logs = None
+    start_time = time.time()
+    interval = 1
+    while True:
+        called_init = await check_run_exists_and_inited(
+            api,
+            entity,
+            project,
+            run_id,
+            run_queue_item_id,
+        )
+        if called_init or time.time() - start_time > RUN_INFO_GRACE_PERIOD:
+            break
+        if not called_init:
+            # Fetch the logs now if we don't get run info on the
+            # first try, in case the logs are cleaned from the runner
+            # environment (e.g. k8s) during the run info grace period.
+            if interval == 1:
+                logs = await run.get_logs()
+            await asyncio.sleep(interval)
+            interval *= 2
+    return called_init, logs
+
+
+async def check_run_exists_and_inited(
+    api: Api, entity: str, project: str, run_id: str, rqi_id: str
+) -> bool:
+    """Checks the state of the run to ensure it has been inited. Note this will not behave well with resuming."""
+    # Checks the _wandb key in the run config for the run queue item id. If it exists, the
+    # submitted run definitely called init. Falls back to checking state of run.
+    # TODO: handle resuming runs
+
+    # Sweep runs exist but are in pending state, normal launch runs won't exist
+    # so will raise a CommError.
+    try:
+        get_run_state = event_loop_thread_exec(api.get_run_state)
+        run_state = await get_run_state(entity, project, run_id)
+        if run_state.lower() != "pending":
+            return True
+    except CommError:
+        wandb.termwarn(
+            f"Run {entity}/{project}/{run_id} with rqi id: {rqi_id} did not have associated run"
```

Comment:
This error message does not appear to match the condition we are testing for.  Doesn't `CommError` imply we couldn't reach the backend?

Reply:
IIRC: We return a CommError when we 404 which is the error we get here. This logic is a copy of the current logic used in our current checks